src/types.rs

```diff
@@ -35,6 +35,14 @@ pub enum ErrorKind {
     InvalidEcRecoverSignature,
 }
 
+/// Errors involving the nonce
+pub enum NonceError {
+    /// Attempted to increment the nonce, but overflow occurred
+    NonceOverflow,
+    /// Account nonce did not match the transaction nonce
+    IncorrectNonce,
+}
+
```

Comment:
I'm personally a big fan of keeping errors limited to the struct they are in. So in this case, it might be better to change all errors to an `enum` called `EngineError` to generalise it instead of having quite a bit of boutique errors.

Reply:
I like when a function's return types are constrained. I think it makes reasoning about the possible errors easier because I don't first need to decide which enum variants are even possible to be returned.

Upstream where error types need to be combined you can always nest the enums

```rust
enum SuperError {
    SubError1(crate::sub_error1::Error),
    // ...
}
```

I understand that nested enums can be a little painful to match on, so another option would be to have a mapping into something like the [`ErrorKind` enum](https://github.com/aurora-is-near/aurora-engine/blob/master/src/types.rs#L31) (a flat, simple structure) we already have

```rust
impl From<SpecificError> for ErrorKind {
    // ...
}
```

All that said, I don't have too strong of a feeling about this, so if you feel very strongly about not having specific error enums, then I will just add variants to [`ErrorKind`](https://github.com/aurora-is-near/aurora-engine/blob/master/src/types.rs#L31) and use that.

Reply:
Sure yes, that was the other part of that, having enums wrapping more granular errors. They are a bit of a pain, it is a bit of boilerplate, but its a good structure.

Reply:
As per discussion, since it is separate, internal, and what not, sure its fine to leave as is.